internal/factsengine/gatherers/sapservices.go

```diff
@@ -0,0 +1,180 @@
+	if len(matches) != 2 {
+		return ""
+	}
+	return matches[1]
+}
+
+type SapServices struct {
+	fs               afero.Fs
+	servicesFilePath string
+}
+
+func NewSapServicesGatherer(servicesFilePath string, fs afero.Fs) *SapServices {
+	return &SapServices{
+		servicesFilePath: servicesFilePath,
+		fs:               fs,
+	}
+}
+
+func NewDefaultSapServicesGatherer() *SapServices {
+	return &SapServices{servicesFilePath: sapServicesDefaultPath, fs: afero.NewOsFs()}
+}
+
+func (s *SapServices) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := []entities.Fact{}
+	log.Infof("Starting %s facts gathering process", SapServicesGathererName)
+
+	entries, err := s.getSapServicesFileEntries()
+	if err != nil {
+		return nil, err
+	}
+
+	factValues, err := convertSapServicesEntriesToFactValue(entries)
+	if err != nil {
+		return nil, SapServicesParsingError.Wrap(err.Error())
+	}
+
+	for _, requestedFact := range factsRequests {
+		fact := entities.NewFactGatheredWithRequest(requestedFact, factValues)
+
+		facts = append(facts, fact)
+	}
+
+	log.Infof("Requested %s facts gathered", SapServicesGathererName)
+	return facts, nil
+}
+
+func (s *SapServices) getSapServicesFileEntries() ([]SapServicesEntry, error) {
+	f, err := s.fs.Open(s.servicesFilePath)
+	if err != nil {
+		return nil, SapServicesFileError.Wrap(err.Error())
+	}
+
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	fileScanner := bufio.NewScanner(f)
+	fileScanner.Split(bufio.ScanLines)
+
+	var entries []SapServicesEntry
+
+	for fileScanner.Scan() {
+		scannedLine := fileScanner.Text()
+		if strings.HasPrefix(scannedLine, "#") || scannedLine == "" {
+			continue
+		}
+		entry := SapServicesEntry{}
+		entry.Content = scannedLine
+
+		if systemdStartup(scannedLine) {
+			entry.Kind = SapServicesSystemdStartup
+			sid := extractSIDFromSystemdService(scannedLine)
+			if sid == "" {
+				return nil, SapServicesParsingError.Wrap(
+					fmt.Sprintf("could not extract sid from systemd sap services entry: %s", scannedLine),
+				)
+			}
+			entry.SID = sid
+		}
+
+		if sapstartStartup(scannedLine) {
+			entry.Kind = SapServicesSapstartStartup
+			sid := extractSIDFromSapstartService(scannedLine)
+			if sid == "" {
+				return nil, SapServicesParsingError.Wrap(
+					fmt.Sprintf("could not extract sid from sapstartsrv sap services entry: %s", scannedLine),
+				)
+			}
+			entry.SID = sid
+		}
+
+		if entry.Kind == "" {
+			// the line is not a recognized entry
+			continue
+		}
+
+		entries = append(entries, entry)
+	}
+
+	return entries, nil
+}
+
+func convertSapServicesEntriesToFactValue(entries []SapServicesEntry) (entities.FactValue, error) {
+	marshalled, err := json.Marshal(&entries)
+	if err != nil {
+		return nil, err
+	}
+
+	var unmarshalled []interface{}
+	err = json.Unmarshal(marshalled, &unmarshalled)
+	if err != nil {
+		return nil, err
+	}
+
+	return entities.NewFactValue(unmarshalled, entities.WithStringConversion())
```

Comment:
I don't think we need `WithStringConversion`, after all, all the values are strings hehe

Reply:
you're right